### ISSUE TITLE
Different behavior of lazy.reject.first

### ISSUE THREAD

**Issue by coorasse** [oracle/truffleruby#1557]
Hello 👋,
After the nice presentation of @eregon today, I was trying to run the tests for [cancancan](https://github.com/CanCanCommunity/cancancan) and they are failing. I could bring down the problem to the following script:
```ruby
magic_array = 10.times.lazy.map do |i|
  raise StandardError.new
end

magic_array.reject(&:nil?).first
puts 'never printed'
```
And here is the output:
```sh
➜  rbenv local 2.6.0
➜  ruby raise.rb
Traceback (most recent call last):
	7: from raise.rb:11:in `<main>'
	6: from raise.rb:11:in `first'
	5: from raise.rb:11:in `each'
	4: from raise.rb:11:in `each'
	3: from raise.rb:11:in `each'
	2: from raise.rb:11:in `each'
	1: from raise.rb:11:in `times'
raise.rb:2:in `block in <main>': StandardError (StandardError)
➜  rbenv local truffleruby-1.0.0-rc10
➜  ruby raise.rb
never printed
```
In ruby 2.6.0 (and previous) the exception is correctly raised, while this is not the case in truffleruby.

**Comment by eregon** [comment]
Thank you for the report!
I have a fix for it which should be merged soon.

**Comment by eregon** [comment]
Fixed in 1927817c7a7754c5453f6c1d325a1bf43a37aa58.
